src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala

```diff
@@ -178,16 +180,60 @@ object ZipAndJarSourcePathFactory extends ZipAndJarFileLookupFactory {
 
 final class FileBasedCache[T] {
   import java.nio.file.Path
-  private case class Stamp(lastModified: FileTime, fileKey: Object)
+  private case class Stamp(size: Long, signatures: ArrayBuffer[Long], fileKey: Object)
   private val cache = collection.mutable.Map.empty[Seq[Path], (Seq[Stamp], T)]
 
+  /**
+    * Read the CRCs of every zip entry if all of them succeed to be read.
+    *
+    * This method uses `java.util.jar` instead of `java.util.zip` utilities
+    * because the method `getCrc` in [[java.util.zip.ZipEntry]] returns 0 or
+    * -1 values.
+    *
+    * @param jarFile A jar file (which follows the zip specification format).
+    * @return A list of checksums for every zip entry.
+    */
+  def readJarCrcs(jarFile: Path): Option[ArrayBuffer[Long]] = {
+    try {
+      val jf = new java.util.jar.JarFile(jarFile.toFile)
+      try {
+        val entries = jf.entries()
+        val buffer = new ArrayBuffer[Long](32)
+        while (entries.hasMoreElements) buffer.+=(entries.nextElement().getCrc)
+        // It is very likely that a jar with no CRCs entries is ill-generated, assume so
+        if (buffer.isEmpty || buffer.forall(_ == -1L)) None else Some(buffer)
+      } finally jf.close()
+    } catch {
+      case _: IOException => None
+    }
+  }
+
+  /**
+    * Get a cached value or create it afresh if the value is missing or
+    * any of the given paths have been "modified".
+    *
+    * This implementation considers that a file has been modified if:
+    *   1. The file has changed its total size.
+    *   2. The CRCs present in the jar file are not the same (in the same order).
+    *   3. The file key (uniquely identifying a file) is different.
+    *
+    * Note that if there's a IO error while reading the CRCs in the cached paths,
+    * this implementation fallbacks to the use of the last modified time. This
+    * also happens when jars have no CRCs -- it is extremely unlikely that a
+    * classpath entry has no CRC, so if that happens we assume the jar is
+    * either corrupt or ill-generated, and we use a safer change detection strategy.
+    *
+    * @param paths The classpath entries to check for modification.
+    * @param create The closure to produce elements of type [[T]].
+    * @return A result of type [[T]] that may be either recently generated or cached.
+    */
   def getOrCreate(paths: Seq[Path], create: () => T): T = cache.synchronized {
     val stamps = paths.map { path =>
       val attrs = Files.readAttributes(path, classOf[BasicFileAttributes])
-      val lastModified = attrs.lastModifiedTime()
-      // only null on some platforms, but that's okay, we just use the last modified timestamp as our stamp
-      val fileKey = attrs.fileKey()
-      Stamp(lastModified, fileKey)
+      val currentSize = attrs.size()
+      val signatures = readJarCrcs(path).getOrElse(ArrayBuffer(attrs.lastModifiedTime().toMillis))
```

Comment:
last modified time should be always kept. The typical use case is that the file is updated when changed, so the optimal comparison is to include that early

Reply:
> last modified time should be always kept. 
  
The reason to remove it from the default strategy is that if I update a jar, I want scalac to look at the checksums of the file and use that as the reliable information to invalidate, not the last modified time, which can change even though the contents are the same (and if it changes, it means we pay a high price for it).

Reply:
What is the use case where the content changes, but we are not affected by the change?

The code currently scans the CRCs even if the modification date is not changed, so we have the code of the open and scan at all times. If there is a  level detection (modification time, then maybe content) we still need to maintain the modification time

Reply:
My goal with this PR is that the modification time should only be a safe fallback, it should not be part of the default strategy, so that other build tools like Bazel and Pants work (hese build tools don't change the modification time at all, they only change contents).
  
As you note, we could probably still have last modification time in the default strategy and have it work with the use cases of these build tools, but I'd like this change to be more robust and support cases in which the contents don't change, but the modification time does -- and still have scalac continue using the entry in the cache.
  
I'm happy to give up on that and add the last modified time if that sounds reasonable to you, but in this case I was just protecting ourselves from the worst-case scenario where, if for some reason the modification time changes and contents don't, we don't trigger JIT deopts.

Reply:
@jvican my primary concern is when I have hundreds of projects each with hundreds or thousands jars on a classpath. I am concerned about being able to compile in reasonable time. Nearly all of these projects use plugins and macros, so I was looking for a net improvement in compiler performance from the caching of plugins and macro classpath

As I stated in the original discussion I do not believe that the the default policy should be bending to specific build tools that deliberately distort their metadata. By extension you cannot trust the CRCs, whether cleared or set to some other value is just the current implementation of the build tools that you have come across. Surely it is up to those tools to provide sufficient metadata, or a plugin to enable a build system to work with them, or we should be super-safe and not cache anything

In https://github.com/scala/scala/pull/6314#issuecomment-365425322 you said (with my **highlighting**) 
> We're on the same page, and your last proposition is the one I'm trying to push for because I believe that asking Bazel to keep that metadata is not a plausible solution; if the CRCs inside the zip file have been computed out of the contents of a zip entry that contained some kind of metadata or filestamps and Bazel removes that metadata, it needs to update the CRC checksums and that is unnecessary IO IMO. **The easiest is to let build tools share fingerprints with the compiler**.

This change, nor the previous change provides this hook the build tool

This change contains an implementation (that we are discussing) but no specification. Moreover the original discussion (https://github.com/scala/scala/pull/6314) did not contain a specification or even an agreement on the path to be taken

If we can install a simple (and extensible) policy (probably path based) then you can ensure that for example for a given classpath and file path
1. you can scan every jar CRC if you want to (like this PR current implementation)
1. you can know that no jar will change (CI environment)
1. you could look at modification dates (if that is good enough for your build environment, or IDE)
1. you could cache nothing
1. you could cache jars in certain directories, without even checking for updates ( e.g. maven repos, ivy repos, company wide managed distributions, JDK etc) 
1. some link to an external build tool that provides information on change when the metadata is not sufficient or performant  
1. etc

This could be just a chain of partial matches and a default, something simple like
```
// just code for the path elements, not the class/source path, classloader cache etc

  trait PathMatchPolicy {
    def policyFor(path:Path): Policy
  }
  class PrefixPathMatchPolicy(prefixes: Seq[Path], policy: Policy, fallback: PathMatchPolicy) extends PathMatchPolicy {
    def policyFor(path: Path): Policy = {
      val abs = path.toAbsolutePath
      if (prefixes.exists{abs.startsWith(_)}) policy
      else fallback(path)
    }
  }
  class JarPathMatchPolicy(policy: Policy, fallback: PathMatchPolicy) extends PathMatchPolicy {
    def policyFor(path: Path): Policy = {
      if (path.endsWith(".jar")) policy
      else fallback(path)
    }
  }
  class DirPathMatchPolicy(policy: Policy, fallback: PathMatchPolicy) extends PathMatchPolicy {
    def policyFor(path: Path): Policy = {
      if (Files.isDirectory(path)) policy
      else fallback(path)
    }
  }
  class FixedPolicy(policy: Policy) extends PathMatchPolicy {
    def policyFor(path: Path): Policy = policy
  }
  // + additional as required

  trait Policy {
    //do the actual caching/lookup
  }
  object CacheNothing extends Policy
  object CacheEverything extends Policy
  object CacheBasedOnFileMetatdata extends Policy
  object CacheBasedOnJarCRC extends Policy
  // + additional as required
```